File: beginner/python/staircase.py
def staircase(n):
  for row in range(n):
    stair = ''
    for col in range(n):
      if (row >= col):
        stair += '#'
      else:
        stair += ' '
    print(stair)

def staircaseRev(n):
  for row in range(n):
    stair = ''
    for col in range(n-1, -1, -1):
      if row >= col:
        stair += '#'
      else:
        stair += ' '
    print(stair)

# staircase recursive function
# def staircase(n, row = 0, stair = ''):
#   if n == row:
#     return

#   if len(stair) == n:
#     print(stair)
#     return staircase(n, row + 1)

#   if len(stair) <= row:
#     stair += '#'
#   else:
#     stair += ' '
  
#   return staircase(n, row, stair)

staircaseRev(9)

File: beginner/javascript/staircase.js
function staircase(n) {
  for (let row = 0; row < n; row++) {
    let stair = '';
    for (let col = 0; col < n; col++) {
      if (row >= col) {
        stair += '#'
      } else {
        stair += ' '
      }
    }
    console.log(stair)
  }
}

function staircaseRev(n) {
  for (let row = 0; row < n; row++) {
    let stair = '';
    for (let col = n-1; col >= 0; col--) {
      if (row < col) {
        stair += ' '
      } else {
        stair += '#'
      }
    }
    console.log(stair)
  }
}

// recursive staircase function, with reverse option
// function staircase(reverse=false, n, row=reverse ? n-1 : 0, stair='') {
//   // exit clause
//   const exitClause = reverse ? row < 0: row === n;
//   if (exitClause) return;

//   // row clause - increment row - fresh stair
//   if (stair.length === n) {
//     console.log(stair);
//     return staircase(reverse, n, reverse ? row-1 : row+1);
//   }

//   // stair clause
//   if (stair.length <= (reverse ? row-1 : row)) {
//     stair += (reverse ? ' ' : '#');
//   } else {
//     stair += reverse ? '#' : ' ';
//   }

//   return staircase(reverse, n, row, stair);
// }

staircaseRev(9)
